pytorch_lightning/utilities/fetching.py

```diff
@@ -169,40 +198,207 @@ def reset(self) -> None:
+            while self.batches:
+                try:
+                    yield_batch = self.pop_batch()
+                    self._fetch_next_batch()
 
-            yield from self._consume_prefetched_batches()
+                    # yield last and has next
+                    self.wait()
 
-    def _consume_prefetched_batches(self) -> Generator:
-        self.done = True
-        while self.batches:
-            if len(self.batches) == 1:
-                yield self.batches.pop(0), True
-            else:
-                yield self.batches.pop(0), False
+                    yield (self.move_data_to_device(yield_batch) if not self.store_on_gpu else yield_batch, False)
+                except StopIteration:
+                    self.batches.insert(0, yield_batch)
+                    break
+
+            yield from self._consume_prefetched_batches()
 
     def _prefetching(self, prefetch_batches: int) -> None:
         for _ in range(prefetch_batches):
             try:
-                batch = next(self.dataloader_iter)
-                self.fetched += 1
-                self.add_batch(batch)
+                self._fetch_next_batch()
             except StopIteration:
                 break
+
+    @contextmanager
+    def apply_profiler(self, name: str) -> Generator:
+        if self.profiler:
+            with self.profiler.profile(name):
+                yield
+        else:
+            yield
+
+    def _fetch_next_batch(self):
+        with self.apply_profiler(f"get_{self.stage}_batch"):
+            with self.fetching_context():
+                data = self.on_fetch_start()
+                with self.apply_profiler(f"fetch_next_{self.stage}_batch"):
+                    batch = next(self.dataloader_iter)
+                self.fetched += 1
+                self.on_fetch_end(batch, data)
+
+    def _consume_prefetched_batches(self) -> Generator:
+        self.done = True
+        while self.batches:
+            yield from self._yield_batch()
+
+    def _yield_batch(self) -> Generator:
+        self.wait()
+        batch = self.batches.pop(0)
+        if not self.store_on_gpu:
+            batch = self.move_data_to_device(batch)
+        is_last = len(self.batches) == 0
+        yield batch, is_last
+
+    def move_data_to_device(self, batch: Any) -> Any:
+        if self.batch_to_device:
+            with self.apply_profiler(f"move_{self.stage}_batch_to_device"):
+                batch = self.batch_to_device(batch)
+        return batch
+
+
+class InterBatchParallelismDataFetcher(DataFetcher):
+
+    """
+    This class implements inter-batch parallelism, which aims at hiding the latency of host-to-device copy
+    of input batches behind computationally intensive operations.
+
+    Without parallelization:
+
+    batch 0: [HtoD][forward][backward]
+    batch 1:                          [HtoD][forward][backward]
+    batch 2:                                                   [HtoD][forward][backward]
```

Comment:
probably has to be formatted as a code block because the character spacing will not be mono, so this illustration will not work.

Reply:
Done !